### ISSUE TITLE
Possible to make the arraymancer compile with the vcc?

### ISSUE THREAD

**Issue by slangmgh** [mratsim/Arraymancer#243]
I try to make the arraymancer compile with the vcc.
There is two part of the code doesn't compile with the vcc.
1. withMemoryOptimHints
2. log1p and expm1

I change the code a bit, the example can compile, but run result is full of error:
```nim
D:\Tool\Nim\tmp\array>ex01_xor_perceptron_from_scratch.exe
 ** On entry to SGEMM  parameter number  0 had an illegal value
 ** On entry to SGEMM  parameter number  0 had an illegal value
Epoch is:0
Batch id:0
Loss is:0.3190621137619019
 ** On entry to SGEMM  parameter number  0 had an illegal value
 ** On entry to SGEMM  parameter number  0 had an illegal value
 ** On entry to SGEMM  parameter number  0 had an illegal value
 ** On entry to SGEMM  parameter number  0 had an illegal value
 ** On entry to SGEMM  parameter number  0 had an illegal value
Epoch is:0
Batch id:1
Loss is:0.3396746218204498
 ** On entry to SGEMM  parameter number  0 had an illegal value
 ** On entry to SGEMM  parameter number  0 had an illegal value
 ** On entry to SGEMM  parameter number  0 had an illegal value
 ** On entry to SGEMM  parameter number  0 had an illegal value
 ** On entry to SGEMM  parameter number  0 had an illegal value
```

What is possible reason of this problem?

**Comment by mratsim** [comment]
Unfortunately I don't have access to a windows machine nor Visual Studio.

### 1. withMemoryOptimHints

here are the change that will probably make it compile:

Change:
```Nim
template withMemoryOptimHints*() =
  when not defined(js):
    {.pragma: align64, codegenDecl: "$# $# __attribute__((aligned(64)))".}
    {.pragma: restrict, codegenDecl: "$# __restrict__ $#".}
  else:
    {.pragma: align64.}
    {.pragma: restrict.}
```

to

```Nim
template withMemoryOptimHints*() =
  when not defined(js):
    {.pragma: restrict, codegenDecl: "$# __restrict $#".}
    when defined(vcc):
      {.pragma: align64, codegenDecl: "$# $# __declspec(align(64))".}
    else: # GCC/Clang
      {.pragma: align64, codegenDecl: "$# $# __attribute__((aligned(64)))".}
    
  else:
    {.pragma: align64.}
    {.pragma: restrict.}
```

I can't test them though.

### 2. `ln1p` and `exp1m`

I don't understand why that would fail, I'm importing from `<math.h>` [like the Nim math module is doing](https://github.com/nim-lang/Nim/blob/855956bf617f68ac0be3717329e9e1181e5dc0c6/lib/pure/math.nim#L121).

I've noticed that Visual Studio requires <cmath.h> instead of <math.h> if you compile for C++ though - https://msdn.microsoft.com/en-us/library/4hwaceh6.aspx. Is that your issue?

Nim math would also be completely broken in that case anyway and the only reason to compile Arraymancer with C++ would be for Cuda (which uses GCC).

### 3. Parameter with illegal value

The error is coming from this call:

https://github.com/mratsim/Arraymancer/blob/ccbd471975b8aa6b116f19635b2c297929e847b1/src/tensor/private/p_operator_blas_l2l3.nim#L156-L161

which calls nimblas and `cblas_dgemm` your machine blas implementation:
https://github.com/unicredit/nimblas/blob/master/nimblas.nim#L60-L65

```Nim
proc gemm*(ORDER: OrderType, TRANSA, TRANSB: TransposeType, M, N, K: int, ALPHA: float32,
  A: ptr float32, LDA: int, B: ptr float32, LDB: int, BETA: float32, C: ptr float32, LDC: int)
  {. dynlib: libName, importc: "cblas_sgemm" .}
proc gemm*(ORDER: OrderType, TRANSA, TRANSB: TransposeType, M, N, K: int, ALPHA: float64,
  A: ptr float64, LDA: int, B: ptr float64, LDB: int, BETA: float64, C: ptr float64, LDC: int)
  {. dynlib: libName, importc: "cblas_dgemm" .}
```

`** On entry to SGEMM  parameter number  0 had an illegal value` is an error reported by Fortran BLAS, I'm not really sure what parameter 0 is (Fortran libraries start at 1).

Apparently linking to gfortran/OpenBLAS are also problematic with Visual Studio - https://github.com/xianyi/OpenBLAS/issues/1437 and OpenBLAS does not seem to be tested with MSVC as well.

**Comment by slangmgh** [comment]
@mratsim 
Thank you. I will check the openblas dll later.

**Comment by slangmgh** [comment]
I trace the assemble code, found that the parameter pass to the function cblas_sgemm is not always correct.
```nim
001AF5A8   00B8D8E2  RETURN to ex01_xor.00B8D8E2 from libopenb.cblas_sgemm
001AF5AC   00000065
001AF5B0   001AF96F
001AF5B4   0000006F
001AF5B8   00000020
001AF5BC   00000003
001AF5C0   00000001
001AF5C4   3F800000
```

For example, the second parameter should be 0000006F, rather than 0x1AF96F, it should be treat as byte, not int32.

Here is the cblas_sgemm:
```nim
cblas_sgem>  55                PUSH    EBP
6D7C9C31     89E5              MOV     EBP, ESP
6D7C9C33     57                PUSH    EDI
6D7C9C34     56                PUSH    ESI
6D7C9C35     53                PUSH    EBX
6D7C9C36     83E4 F0           AND     ESP, 0xFFFFFFF0
6D7C9C39     83EC 70           SUB     ESP, 0x70
6D7C9C3C     8B55 08           MOV     EDX, [EBP+0x8]
6D7C9C3F     8D4D 20           LEA     ECX, [EBP+0x20]
6D7C9C42     8B45 1C           MOV     EAX, [EBP+0x1C]
6D7C9C45     894C24 44         MOV     [ESP+0x44], ECX
6D7C9C49     8D4D 34           LEA     ECX, [EBP+0x34]
6D7C9C4C     894C24 48         MOV     [ESP+0x48], ECX
6D7C9C50     83FA 66           CMP     EDX, 0x66
6D7C9C53     C74424 30 0000000>MOV     DWORD PTR [ESP+0x30], 0x0
6D7C9C5B     0F84 11010000     JE      libopenb.6D7C9D72
6D7C9C61     83FA 65           CMP     EDX, 0x65
6D7C9C64     74 2A             JE      SHORT libopenb.6D7C9C90
6D7C9C66     8D4424 30         LEA     EAX, [ESP+0x30]
6D7C9C6A     C74424 08 0700000>MOV     DWORD PTR [ESP+0x8], 0x7
6D7C9C72     894424 04         MOV     [ESP+0x4], EAX
6D7C9C76     C70424 C0188F6E   MOV     DWORD PTR [ESP], libopenb.6E8F18C0                     ; ASCII "SGEMM "
6D7C9C7D     E8 EE282100       CALL    libopenb.xerbla_
6D7C9C82     8D65 F4           LEA     ESP, [EBP-0xC]
6D7C9C85     5B                POP     EBX
6D7C9C86     5E                POP     ESI
6D7C9C87     5F                POP     EDI
6D7C9C88     5D                POP     EBP
6D7C9C89     C3                RETN
```

**Comment by slangmgh** [comment]
I think importc with Enum type is the problem.
When I change gemm define to:
```nim
proc gemm*(ORDER: int, TRANSA, TRANSB: int, M, N, K: int, ALPHA: float32,
  A: ptr float32, LDA: int, B: ptr float32, LDB: int, BETA: float32, C: ptr float32, LDC: int)
  {. dynlib: libName, importc: "cblas_sgemm" .}
```

Then the call is ok.

**Comment by slangmgh** [comment]
As for ln1p, seems there is something strange with the VC 2013, when I use the lib from vc, the code output lots of "Loss is: nan", so I add the following code to replace the ln1p:
```nim
proc ln1p*[T: float32|float64](x: T): T =
  var y, z: T
  y = 1 + x
  z = y - 1
  return ln(y) - (z - x) / y
```

Now the program run fine.

**Comment by mratsim** [comment]
Nice thanks for the investigation.

Probably the issue is in @andreaferretti nimblas.

I suppose [this here](https://github.com/unicredit/nimblas/blob/master/nimblas.nim#L17-L21)

```Nim
type
  TransposeType* = enum
    noTranspose = 111, transpose = 112, conjTranspose = 113
  OrderType* = enum
    rowMajor = 101, colMajor = 102
```

Should be
```
type
  TransposeType* {.size: sizeof(cint).}= enum
    noTranspose = 111, transpose = 112, conjTranspose = 113
  OrderType* {.size: sizeof(cint).}= enum
    rowMajor = 101, colMajor = 102
```

For ln1p, it's strange mmh.

**Comment by andreaferretti** [comment]
@mratsim Thank you, I will fix it ASAP (tomorrow at worst)

**Comment by slangmgh** [comment]
@mratsim @andreaferretti 
Should be 
```nim
TransposeType* {.size: sizeof(int).}= enum
```

**Comment by andreaferretti** [comment]
Fixed in nimblas 0.2.1
